_posts/2020-11-10-jit-part-2.md

```diff
@@ -0,0 +1,192 @@
+---
+layout: post
+title: JIT part 2
+tags: BEAM erts jit
+author: John Högberg
+---
+
+This post continues our [adventures in the JIT], digging a bit deeper into the
+implementation details.
+
+While writing things in machine code (assembler) gives you great freedom, it
+comes at the cost of having to invent almost everything yourself. For example,
+if you call a function in a certain manner and said function doesn't expect
+that, you'll crash your OS process at best or spend hours chasing a [heisenbug]
+at worst.
+
+Hence, _conventions_ are always front and center when writing assembler, so we
+need to visit some of the ones we've chosen before moving on.
+
+The most important one concerns registers, and we base it on the system calling
+convention to make it easier to call C code. I've included tables for the
+SystemV convention used on Linux below. The registers differ on other systems
+like Windows, but the principle is the same on all of them.
+
+Register | Name | Callee save | Purpose
+| :- | :-: |:-: | -: |
+`RDI` | `ARG1` | no | First argument
+`RSI` | `ARG2` | no
+`RDX` | `ARG3` | no
+`RCX` | `ARG4` | no
+`R8` | `ARG5` | no
+`R9` | `ARG6` | no | Sixth argument
+`RAX` | `RET` | no | Function return value
+
+Thus, if we want to call a C function with two arguments, we move the first
+into `ARG1` and the second into `ARG2` before calling it, and we'll have the
+result in `RET` when it returns.
+
+Beyond saying which registers are used to pass arguments, calling conventions
+also say which registers retain their value over function calls. These are
+called "callee save" registers, since the called function needs to save and
+restore them if they're modified.
+
+In these registers, we keep commonly-used data that rarely (if ever) changes in
+C code, helping us avoid saving and restoring them whenever we call C code:
+
+Register | Name | Callee save | Purpose
+| :- | :-: |:-: | -: |
+`RBP` | `active_code_ix` | yes | Active code index
+`R13` | `c_p` | yes | Current process
+`R15` | `HTOP` | yes | Top of the current process' heap
+`R14` | `FCALLS` | yes | Reduction counter
+`RBX` | `registers` | yes | BEAM register structure
+
+We also keep the current process' stack in `RSP`, the _machine stack pointer_,
+to allow `call` and `ret` instructions in Erlang code.
+
+The downside of this is that we can no longer call arbitrary C code as it may
+assume a much larger stack, requiring us to swap back and forth between a "C
+stack" and the "Erlang stack".
+
+In my previous post we called a C function (`timeout`) without doing any of
+this, which was a bit of a white lie. It used to be done that way before we
+changed how the stack works, but it's still pretty simple as you can see below:
+
+```c++
```

Comment:
Add a link to the code?

Reply:
Do you mean `timeout` or the code in the previous post?

Reply:
timeout